### PR TITLE
Export Gradient Map from MMDLoader to let it reusable

### PR DESCRIPTION
**Pull request by takahirox**
I exported toon map cel shading technique from `MMDLoader` to `MeshPhongMaterial`
to make it reusable for non-MMD.

With `toonMap` and `OutlintEffect` you can easily render toon style objects on `Three.js`.

Left: Cel Shading On (`toonMap` + `OutlineEffect`)
Right: Cel Shading Off

image
(Model: Tda's Miku append http://seiga.nicovideo.jp/seiga/im2018614)

How to use:

```
var material = new THREE.MeshPhongMaterial();
material.toonMap = new THREE.Texture( toonImage );
material.toonMapDirectionY = <bool>; // default is false
```

toonImage should be like this

image

The technique this PR includes is a very basic cel shading technique.

```
float dotNL = dot( geometry.normal, directLight.direction );
vec2 coord = vec2( dotNL * 0.5 + 0.5, 0.0 );
return texture2D( toonMap, coord ).rgb;
```

This RGB color will be used for calculating irradiance.
This is the trick which makes objects look like flat.

In general, direction X in `toonImage` is used for that 
but sometimes direction Y is used, for example in MMD.
So I enabled users to specify it with `toonMapDirectionY`.

Note:
specular can be more cel-ish but I'm not sure what technique is the most standard.
It'll be a future work.

### CODE REVIEW

File: src/materials/MeshPhongMaterial.js [hunk 1]
```diff
@@ -94,6 +97,9 @@ function MeshPhongMaterial( parameters ) {
 	this.reflectivity = 1;
 	this.refractionRatio = 0.98;
 
+	this.toonMap = null;
```
Comment (mrdoob) [inline src/materials/MeshPhongMaterial.js]: Hmm, what's a `toonMap`?

Comment (takahirox) [inline src/materials/MeshPhongMaterial.js]: Reference: http://gamedev.stackexchange.com/questions/51063/what-is-ramp-shading-or-lighting

`rampMap` would be better than `toonMap`?

Comment (mrdoob) [inline src/materials/MeshPhongMaterial.js]: Ah, it's a gradient? And it takes lights into account, right?

Comment (takahirox) [inline src/materials/MeshPhongMaterial.js]: Yes.

With like the toonImage I put above
it can make object look like flat=toon-style.

**Comment by takahirox** [comment]
I've renamed `toonMap` with `gradientMap`.
How does this PR sound to you now?

**Comment by takahirox** [comment]
\ping @mrdoob

**Comment by takahirox** [comment]
BTW, by applying this PR legacy code `UniformsUtils.merge()` and `UniformsUtils.clone()` will be removed from `MMDLoader`.

File: src/materials/MeshPhongMaterial.js [hunk 2]
```diff
@@ -94,6 +97,9 @@ function MeshPhongMaterial( parameters ) {
 	this.reflectivity = 1;
 	this.refractionRatio = 0.98;
 
+	this.gradientMap = null;
+	this.gradientMapAxisY = false;
```
Comment (mrdoob) [inline src/materials/MeshPhongMaterial.js]: Do you mind removing `gradientMapAxisY` and the related code? I think I would prefer "forcing" the X axis gradient.

Comment (takahirox) [inline src/materials/MeshPhongMaterial.js]: Alright!

Comment (mrdoob) [inline src/materials/MeshPhongMaterial.js]: I'm also thinking that, maybe `MeshToonMaterial` could be more appropriate here... 😶

**Comment by takahirox** [comment]
I've removed `gradientMapAxisY`.
About `MeshToonMaterial`, hm let me try.

**Comment by takahirox** [comment]
I've made `MeshToonMaterial`!

**Comment by mrdoob** [comment]
Thanks! Let me double-check with @WestLangley though...

@WestLangley What do you think? Just adding `gradientMap` to `MeshPhongMaterial` or creating `MeshToonMaterial` which extends `MeshPhongMaterial`?

We definitely need to do some `ifdefs` clean up one way or another...

**Comment by WestLangley** [comment]
>@WestLangley What do you think? Just adding gradientMap to MeshPhongMaterial or creating MeshToonMaterial which extends MeshPhongMaterial?

I think creating `MeshToonMaterial` is best.

However, I also think we should refactor and remove the material-case-statements from the renderer; in other words, make the renderer material-agnostic.

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
Seems like you forgot to include `meshtoon_vert.glsl` and `meshtoon_frag.glsl`.

**Comment by takahirox** [comment]
Oops! I'll add now!

**Comment by takahirox** [comment]
Done!

**Pull request merged**